Add more to Changing Python (#1671)
python/devguide: developer-workflow/lang-changes.rst

.. _lang-changes:
.. _langchanges:

Changing Python
===============

On occasion people come up with an idea for changing or improving Python.
This page explains how to explore change proposals and what changes
have a reasonable chance of being considered.

Considerations
--------------

First, understand that changes to Python
are difficult to make. When the language changes,
**every** Python programmer and all Python programmers to
come will end up eventually learning about the change you want to propose.
Books will need updating, code will be changed, and a new way to do things will
need to be learned. Changes to Python are never taken
lightly.

Backward compatibility is a significant concern.  Existing Python code has to
continue to work.  There are exceptions to this rule, but they are very rare
and are only allowed when the benefits of the change greatly outweigh the costs
of breaking existing code.

Because of the seriousness of changing the language, any change must be

> beneficial to a many Python users. If the change only helps a

Reviewer: ```suggestion
beneficial to many Python users. If the change only helps a
```

small percentage of Python developers then the change will not be made. A good
way to see if your idea would work for a large portion of the Python community
is to ask in the `Ideas Discourse category <ideas_>`_. You can also
look in Python's standard library to find examples of code which would benefit from
your proposed change.  This can also help communicate the usefulness of your change to

> others.

Reviewer: ```suggestion
is to ask in the `Ideas Discourse category <ideas_>`_.   This can help
communicate the usefulness of your idea to others.  You can also
look in Python's standard library to find examples of code which would benefit from
your proposed change.
```
I think above is a more natural ordering, looking the the stdlib will not help communicate your idea. I also suggest changing "change" to "idea" for consistency.

Author: I removed the sentence about communicating the usefulness of the idea, that's covered below.

For all of these reasons, most proposed changes to Python are rejected.  This
doesn't mean you shouldn't suggest them.  It can be useful to explore
alternatives and to get feedback from the community.  Just be aware that

> getting a change accepted is difficult.  Don't take it personally.

Author: Done

.. _lang-changes:
.. _langchanges:

Changing Python
===============

On occasion people come up with an idea for changing or improving Python.  This
page explains how to propose changes and what to expect during the process.

Considerations
--------------

First, understand that changes to Python are difficult to make. When the
language changes, **every** Python programmer and all Python programmers to
come will end up eventually learning about the change you want to propose.
Books will need updating, code will be changed, and a new way to do things will
need to be learned. Changes to Python are never taken lightly.

> Backward compatibility is a significant concern.  Existing Python code has to

Reviewer 2: ```suggestion
Backward compatibility is a significant concern. Existing Python code has to
```

> continue to work.  There are exceptions to this rule, but they are very rare

Author: No need to edit these all manually.

and are only allowed when the benefits of the change greatly outweigh the costs
of breaking existing code.

Because of the seriousness of changing the language, any change must be
beneficial to many Python users. If the change only helps a small percentage of
Python developers then the change will not be made. A good way to see if your
idea would work for a large portion of the Python community is to discuss it in
the `Ideas category in Discourse <ideas_>`_.  You can also look in Python's
standard library to find examples of code which would benefit from your
proposed change.

.. important::
   For all of these reasons, most proposed changes to Python are rejected.  This
   doesn't mean you shouldn't suggest them.  It can be useful to explore
   alternatives and to get feedback from the community.  Just be aware that

>    getting a change accepted is difficult.  Don't take it personally.

Reviewer 2: ```suggestion
   For all of these reasons, most proposed changes to Python are rejected. This
   doesn't mean you shouldn't suggest them. It can be useful to explore
   alternatives and to get feedback from the community. Just be aware that
   getting a change accepted is difficult. Don't take it personally.
```

Author: I'll remove the double spaces......

.. _lang-changes:
.. _langchanges:

Changing Python
===============

On occasion people come up with an idea for changing or improving Python.
This page explains how to propose changes and what to expect during the

> process.

Author: Added.  It's a little broader that just the language, but I take your point.

Suggesting new features and language changes
--------------------------------------------

Proposing a change involves the following steps:

- Describe your idea in detail.

- Engage in discussion about the idea.

- Rarely, the idea will advance to a formal proposal stage.

Describe your idea
^^^^^^^^^^^^^^^^^^

The `Ideas category in Discourse <ideas_>`_ is specifically intended for
discussion of new features and language changes. Make your proposal as a `new
topic <ideas_>`_ there.

Your proposal needs to be **detailed**. Describe the change you want to make,
why you want to make it, and what benefits it will bring to Python users. Be
specific. Show that you have considered the effect of the change on existing
code.

Some things you should **not** do:

- **Don't** format your idea as a Python Enhancement Proposal (*PEP*).
  It's good to use the PEP template as a
  guide for what information to include, but don't try to write a full PEP

>   until the idea has been accepted for further consideration.

Author: I'm not sure how best to describe the right balance between brevity and detail.

>   until the idea has been accepted for further consideration.

Reviewer 2: As a starter for ten?:

> "Ensure your description covers all the salient points of the proposal, and make it as short as possible, but no shorter. This maximises the liklihood that other members of the community will have the time to read and consider your proposal, and is a considerate way to be respectful of their time"

>   until the idea has been accepted for further consideration.

Author: I added a bullet to the "Do" list about balancing brevity and detail.

Some things you **should** do:

- Research whether your idea has been proposed before. There are many
  suggestions that have been made and rejected in the past. If your idea has a
  history, read the discussion to see why it was rejected. Previously rejected
  ideas are especially unlikely to get accepted. If you can address the
  concerns raised in the previous discussion, mention that in your proposal.

- Read other ideas to see how they are presented. This will help you flesh out
  your proposal.

- Read other ideas to understand the kinds of concerns that are raised and
  objections that must be answered. Your proposal will be stronger if you can
  address these concerns up front.

Engage in discussion
^^^^^^^^^^^^^^^^^^^^

Once you post your proposal, people will respond. You need to continue
the discussion, answer questions, and address objections. This is an important
part of the process. Other people's perspectives will help explore the full
impact of the idea and find strengths or weaknesses you may not have
considered.

You will almost always get at least some push-back.  Don't be disappointed, and don't take it personally.

> As the long list of :pep:`Withdrawn and Rejected PEPs

Author: I'll reflow all the paragraphs before merging.

--------------------------------------------

Proposing a change involves the following steps:

- Describe your idea in detail.

- Engage in discussion about the idea.

- Rarely, the idea will advance to a formal proposal stage.

Describe your idea
^^^^^^^^^^^^^^^^^^

The `Ideas category in Discourse <ideas_>`_ is specifically intended for discussion
of new features and language changes. Make your proposal as a `new topic
<ideas_>`_ there.

Your proposal needs to be **detailed**. Describe the change you want to make,
why you want to make it, and what benefits it will bring to Python users. Be
specific. Show that you have considered the effect of the change on existing
code.

Some things you should **not** do:

- **Don't** format your idea as a Python Enhancement Proposal (*PEP*).
  It's good to use the PEP template as a guide for what information to include,
  but don't try to write a full PEP until the idea has been accepted for
  further consideration.

- **Don't** create a list of proposed changes. Each idea needs its own
  discussion. Proposing a change is significant work. You will need to spend
  time on each proposal. A tossed-off list will not be taken seriously.

> When justifying your idea, these are not good reasons to make a change:

Reviewer 3: ```suggestion
When justifying your idea, these are **not good reasons** to make a change:
```
To highlight the negateion as in the section before.

Author: Done.

  discuss, so long as it includes all the necessary information.

Engage in discussion
^^^^^^^^^^^^^^^^^^^^

Once you post your proposal, people will respond. You need to continue
the discussion, answer questions, and address objections. This is an important
part of the process. Other people's perspectives will help explore the full
impact of the idea and find strengths or weaknesses you may not have
considered.

You will almost always get at least some push-back. Don't be disappointed,
and don't take it personally.
As the long list of :pep:`Withdrawn and Rejected PEPs
<0#rejected-superseded-and-withdrawn-peps>` in the :pep:`PEP Index <0>`
attests, and as befits a mature programming language, getting significant
changes into Python isn't a simple task.

Engage in good faith. The goal of the discussion is to find the best balance
between competing concerns. If your idea has merit, the discussion will help
refine it and make it stronger. If your idea has flaws, the discussion will
help identify them so you can address them or move on. Keep the discussion
productive and focused on the issues, not on personalities.

> It's especially useful to get core team members involved in the discussion.

Reviewer 3: Can this be made more specific? Like

> It is especially useful to discuss with core team members because
> they know the language and boundary conditions very well,
> and eventually you'll need a core team member to sponsor a
> potential PEP.

"get core team members involved" may be misunderstood as actively asking. I've seen pings to core team members to get attention, but that's certianly not what we want to encourage.

Author: I added similar language.

Suggesting new features and language changes
--------------------------------------------

Proposing a change involves the following steps:

- Describe your idea in detail.

- Engage in discussion about the idea.

- Rarely, the idea will advance to a formal proposal stage.

Describe your idea
^^^^^^^^^^^^^^^^^^

The `Ideas category in Discourse <ideas_>`_ is specifically intended for discussion
of new features and language changes. Make your proposal as a `new topic
<ideas_>`_ there.

Your proposal needs to be **detailed**. Describe the change you want to make,
why you want to make it, and what benefits it will bring to Python users. Be
specific. Show that you have considered the effect of the change on existing
code.

Some things you should **not** do:

> - **Don't** format your idea as a Python Enhancement Proposal (*PEP*).

Reviewer 4: ```suggestion
- **Don't** format your idea as a Python Enhancement Proposal (PEP).
```

When justifying your idea, these are not good reasons to make a change:

- "Other languages work this way." Languages are different. What works well
  in one language may not work well in Python, or Python may already have a
  way to do a similar thing.

- "It would be nice." This is too vague. Be specific about the benefits
  your change will bring, and consider the costs as well.

Some things you **should** do:

- Research whether your idea has been proposed before. There are many
  suggestions that have been made and rejected in the past. If your idea has a
  history, read the discussion to see why it was rejected. Previously rejected
  ideas are especially unlikely to get accepted. If you can address the
  concerns raised in the previous discussion, mention that in your proposal.

- Read other ideas to see how they are presented. This will help you flesh out
  your proposal.

- Read other ideas to understand the kinds of concerns that are raised and
  objections that must be answered. Your proposal will be stronger if you can
  address these concerns up front.

Engage in Discussion
^^^^^^^^^^^^^^^^^^^^

Once you post your proposal, people will respond. You need to continue

> the discussion, answer questions, and address objections. This is an important

Author: In my experience, every post gets a response, no matter its quality.

part of the process. Other people's perspectives will help explore the full
impact of the idea and find strengths or weaknesses you may not have
considered.

> You will get push-back.  Don't be disappointed, and don't take it personally.

Reviewer 5: "You will almost always get push-back" works for me.

> You will get push-back.  Don't be disappointed, and don't take it personally.

Author: I added my proposed text.

PEP process
-----------

Once you are certain you have a language change proposal which will appeal to
the general Python community, you can begin the PEP
process to officially propose the change.  See :pep:`1` for
information on PEPs and the PEP process, and the :pep:`PEP Index <0>` for

> examples.

Reviewer 2: Why change the wrapping here? Starting sentences/clauses on their own lines reduces future diffs, and can generally be a good idea.

Author: I can't tell what you are suggesting: 1) I should change the line breaks to correspond to clauses, or 2) I shouldn't change the wrapping from the text on main?   Some paragraphs have been re-wrapped already.

> If the PEP is accepted, then your language change will be introduced in a

Reviewer 2: Not all changes are language changes, &c
```suggestion
If the PEP is accepted, then your proposal will be introduced in a
```

PEP process
-----------

Once you are certain you have a language change proposal which will appeal to
the general Python community, you can begin the PEP
process to officially propose the change. See :pep:`1` for
information on PEPs and the PEP process, and the :pep:`PEP Index <0>` for

> examples.

Author: Added.
